stcrpy/tcr_metrics/tcr_dockq.py

```diff
@@ -0,0 +1,405 @@
+        """
+        Initialize the TCRDockQ class. See `Dockq --help` for more information.
+
+        Args:
+            TCR_pMHC_interface (bool): Whether to use evaluate the entire TCR-pMHC interface or evaluate spearate chains. Deaults to True. This will create a merged psudo-chain for the antigen and the TCR.
+            **kwargs: Keyword arguments from DockQ: mapping, small_molecule, capri_peptide, no_align, json, n_cpu, max_chunk, allowed_mismatches, verbose, short, print_results.
+        """
+
+        DockQArgs = namedtuple(
+            "DockQArgs",
+            [
+                "mapping",
+                "small_molecule",
+                "capri_peptide",
+                "no_align",
+                "json",
+                "n_cpu",
+                "max_chunk",
+                "allowed_mismatches",
+                "verbose",
+                "short",
+                "print_results",
+            ],
+        )
+        self.TCR_pMHC_interface = TCR_pMHC_interface
+        self.args = DockQArgs(
+            mapping=kwargs.get("mapping", None),
+            small_molecule=kwargs.get("small_molecule", None),
+            capri_peptide=kwargs.get("capri_peptide", False),
+            no_align=kwargs.get("no_align", False),
+            json=kwargs.get("json", None),
+            n_cpu=kwargs.get("n_cpu", 8),
+            max_chunk=kwargs.get("max_chunk", 512),
+            allowed_mismatches=kwargs.get("allowed_mismatches", 0),
+            verbose=kwargs.get("verbose", False),
+            short=kwargs.get("short", False),
+            print_results=kwargs.get("print_results", False),
+        )
+        return
+
+    def try_tcr_mapping(self, dock: "abTCR", reference: "abTCR"):
+        try:
+            dock_chain_mapping = {v: k for k, v in dock.get_chain_mapping().items()}
+            ref_chain_mapping = {v: k for k, v in reference.get_chain_mapping().items()}
+
+            mapping = [
+                (dock_chain_mapping[chain_type], ref_chain_mapping[chain_type])
+                for chain_type in dock_chain_mapping
+                if chain_type in ref_chain_mapping
+            ]
+            mapping = list(zip(*mapping))
+            string_mapping = f"{''.join(mapping[0])}:{''.join(mapping[1])}"
+            return mapping, string_mapping
+        except Exception:
+            return None, None
+
+    def retrieve_chain(self, tcr: "TCR", chain_id: str):
+        chain_map = tcr.get_chain_mapping()
+        if chain_map[chain_id] in ["VA", "VB", "VD", "VG"]:
+            return tcr[chain_id]
+        elif chain_map[chain_id] == "Ag":
+            return [ag for ag in tcr.get_antigen() if ag.id == chain_id][0]
+        else:
+            return tcr.get_MHC()[0][chain_id]
+
+    def filter_residues(self, dock: "abTCR", reference: "abTCR", mapping):
+        from stcrpy.tcr_formats.tcr_formats import get_sequences
+
+        filtered_dock = copy.deepcopy(dock)
```

Comment:
deepcopy is usually pretty slow. Since the TCR objects inherit from Entity, there should be copy method that can be used that will speed up the calculations. https://biopython.org/docs/1.76/api/Bio.PDB.Entity.html#Bio.PDB.Entity.Entity.copy

Possibly, there are some attributes that would need to be implemented specific to the TCR so it may need to be overwritten

Reply:
Agreed - this is the result of a very odd bug, whereby copy's shallow copy causes problems when iteratively indexing over and detaching residues from the copy. We can review at a later stage if this needs to be optimised at any point.